docs/features/non-source-outputs.md

```diff
@@ -0,0 +1,141 @@
+
+```ini
+compiler_artifact.Microsoft.CodeAnalysis.Razor.SourceGenerators.SourceGenerator.outputdir = bin\net8.0\Razor
+compiler_artifact.ArtifactName.outputdir = c:\dir
+```
+
+This means that generator authors are free to specify a mapping alongside their generator via the standard `NuGet` props and targets mechanism, and consumers of the generators are able to override the location in a consistent way.
+
+## MSBuild semantics
+
+Today, it's not possible to dynamically provide the list of outputs to an `MSBuild` target. This proposal allows the compiler to effectively emit an unbound number of files, none of which are tracked via `MSBuild` meaning that incremental compilation can be broken. (Note this is actually true today thanks to `EmitCompilerGeneratedFiles` but they don't have a bearing on the correctness of build).
+
+In order to correctly participate in the build process the compiler and targets are updated to emit a new file with the list of all files that it wrote as part of a compilation. A new target is added before `CoreCompile` that runs unconditionally and looks for the existence of the output list file; if the list is present and anything in the list is missing or out of date, the target touches a 'sentinel file'. The 'sentinel' file is passed as an input to `CoreCompile`. This ensures that anytime an output from the compiler is modified outside of the compilation process, `CoreCompile` will be correctly invoked and not skipped as being up to date. The 'sentinel' file effectively stands in for all of the outputs produced by the compiler in the `MSBuild` up to date checks.
+
+In order to decide if a file is modified the new target will require a task that can decide if a file has been modified since the previous build. This logic is fairly simple, as it just has to compare the timestamp of the output list file with the items listed within, but it's worth calling out that this does duplicate some of the `MSBuild` logic in the compiler tasks. The concept of dynamic inputs/outputs is proposed by `MSBuild` but not currently implemented. If this feature is added we could remove this duplication.
+
+## Open Questions
+
+- Naming: This proposal originally used the term `NonSource` but that's not particularly descriptive and perhaps odd in that it's not actually the mirror of a source output in that it's not going into the compilation. `Artifact` seems like a better term? Or simply `File`?
+- Do we want custom mapping for the `compiler_artifact.` scheme, or could we just use the existing `build_metadata.<ItemType>.<Metadata>` mechanism and recognize `CompilerMappedArtifact` as a special key?
+  - This is doable, but does tie the MSBuild item type name directly into the compiler, rather than having a level of indirection.
+- File mapping: should we just let the generators party on the file system and allow absolute paths and `.`, `..` etc, rather than having the host control the output location?
```

Comment:
I feel like this is the route that we should be going. Effectively the `AddFile` API should allow for relative or full paths. In the case of relative it's treated like hint source today and emitted in the generator specific directory. If it's a full path then it's just emitted where the path says. 

As you note that does raise the problem of how do generators get information about the project layout. Think we can help them by essentially auto-feeding the project file path into analyzers options. That way there is a point of reference they can use.

This is a bit of a side bar but I've long wondered if there are a few common / important MSBuild properties that we should feed through to analyzer config. Analyzers want to see items like `<TargetFramework>` and having them all feed it through feels like wasted effort. Or worse watching them [parse it out of metadata][1]. This seems like anotherc case where we could solve a few problems by having a well known set of porperties that we auto-feed through analyzer config options. 



[1]: https://github.com/dotnet/roslyn-analyzers/pull/7116

Reply:
I guess allowing the generators to write absolute paths or whatever makes me *really* nervous from an escape hatch point of view. If a generator is writing the wrong place, or I have some weird build setup where I do things post-build, it's painful if I have no control over where the artifacts are going.

I guess it seems weird that everything else the compiler outputs is controllable, but suddenly there is a whole set of stuff that is effectively hard coded by the compiler.

Reply:
> If a generator is writing the wrong place, or I have some weird build setup ...

What is the difference though between writing to the wrong place and copying to the wrong place? Both designs allow for the generator author to, at least initially, control where they write files. Mistake can happen in either place. I'd argue mistakes are more likely to happen when we abstract it such that success means combining locations from different sources like .globalconfig, source code and msbuild mappings. 
 
> where I do things post-build, it's painful if I have no control over where the artifacts are going.

For some generators, like `wwworoot`, giving control effectively breaks the generator. Using it is an understanding that it's going to put the files where it wants. I'm not sure what it would mean to give users control over where generators like this emit output. 

For the _emit to disk and use it later in build_ case I can see that customization is more desirable. But if the default location for such artifacts is under `obj` and users can already control `obj` location ... what is the value in more customization? Definitely get the value in customizing `obj` (done it many many times) but customizing within `obj` i can't see a good use case for (am I missing an important one?). I also worry about the more customizations we add the more ways users can make their builds complex and hard to understand by tooling. So I'm wary of striking the right balance.

Reply:
I think the difference between the path being in MSBuild and the path being in the generator itself, it that as a user I can always override the MSBuild logic. It has natural, understood extension points and ways to overwrite values. If it's in the generator I have zero control, other than removing the generator altogether.

Reply:
What is the value in overriding a generator that can only output correctly to one place?

Reply:
> This is a bit of a side bar but I've long wondered if there are a few common / important MSBuild properties that we should feed through to analyzer config. Analyzers want to see items like `<TargetFramework>` and having them all feed it through feels like wasted effort. Or worse watching them [parse it out of metadata](https://github.com/dotnet/roslyn-analyzers/pull/7116). This seems like anotherc case where we could solve a few problems by having a well known set of porperties that we auto-feed through analyzer config options.

I like this.

Reply:
I get that _you_ want that control. But for the majority of our users MSBuld is a scary place and they are just going to use the default from the generator author. I would like to see some examples of why the user would override the generator author. I feel like the generator author is probably writing code that accesses the artifact, but maybe I am missing some use cases.

Reply:
I guess my biggest worry is that if you allow the generator to decide the output path then you're hard coding layout assumptions into the generator itself. Sure `wwwroot\js` might be the right place for the files today, but what if tomorrow Blazor turns around and moves that directory?